### PR TITLE
Fix all AjaxDataTables bugs and known issues

### PR DESCRIPTION
**Pull request by tabacitu**
This PR is supposed to fix all AjaxDataTables issues, so that we can set is as the default way we show information. The removal of the "classic" datatables is to be done in another branch, in case this DOES solve all our problems.

Let's document all issues and make sure this fixes them. All praise @indra1 for taking the time to remove the "live control" package, which allows us to even hope to accomplish this.

### CODE REVIEW

**Comment by indra1** [comment]
@tabacitu I will update my composer today an see how it works on a large project

**Comment by tabacitu** [comment]
Issues and their status, starting with @OwenMelbz 's roundup.

| Issue        | Description           | Status  |
| ------------- |:-------------:| -----:|
| https://github.com/Laravel-Backpack/CRUD/issues/411      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/402      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/407      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/370      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/357      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/306      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/324      | todo | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/183      | todo | todo |

**Comment by lloy0076** [comment]
| Issue        | Description           | Status  |
| ------------- |:-------------:| -----:|
| https://github.com/Laravel-Backpack/CRUD/issues/411      | Streamline enableAjaxTable with functionality. | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/402      | Concatenate two columns in list view | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/407      | access to all model attributes in button | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/370      | filter by pivot table | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/357      | problem with orderBy and enableAjaxTable  | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/306      | model_function Column with enableAjaxTable | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/324      | Timeout with large datasets | todo |
| https://github.com/Laravel-Backpack/CRUD/issues/183      | Saerch in Table With Relationship | todo |

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 1]
```diff
@@ -8,48 +8,118 @@
...
+     */
+    private function make($data)
+    {
+        $rows = [];
+        foreach ($data as $row) {
+            $rows[] = $this->format($row);
+        }
+
+        return [
+            'draw'            => (isset($this->input['draw']) ? (int) $this->input['draw'] : 0),
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: I'd drop the surrounding `()` breaces - it took me too long to figure out what where the condition was, if it was braced correctly and more.

In fact I'd almost be tempted to refactor the ternary condition out to:

`$shouldDraw = // That awful condition` and use `$shouldDraw` there.

Also that smells like a Boolean and not a 1/0.

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 2]
```diff
@@ -8,48 +8,118 @@
...
+            'recordsTotal'    => $this->totalRows,
+            'recordsFiltered' => $this->filteredRows,
+            'data'            => $rows,
+        ];
+    }
+
+    /**
+     * Checks of teh user has anything written in the search bar and if so a string with the filtered is returned.
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: If my browser keeps on disallowing me from writing **t** **e** **h** I'm going to scream :P Check the spelling :P

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 3]
```diff
@@ -8,48 +8,118 @@
...
+     * In case no filter is detected null is returned.
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: Surround null with backspaces maybe?

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 4]
```diff
@@ -8,48 +8,118 @@
...
+     *
+     * @return null | string $filter
+     */
+    private function addSearchConditions()
+    {
+        if (isset($this->input['search']) && isset($this->input['search']['value'])) {
+            $filter = $this->input['search']['value'];
+            if ($filter !== '') {
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: `! empty($filter) ` maybe?

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 5]
```diff
@@ -8,48 +8,118 @@
...
+                return $filter;
             }
+        }
+    }
 
-            return $row_items;
-        });
+    /**
+     * Checks if the user tried to order a column and if so an array with the
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: There's a verb missing after that and - my guess it's *return* but I could be wrong.

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 6]
```diff
@@ -8,48 +8,118 @@
...
+    {
+        if (isset($this->input['order']) && isset($this->input['order'][0])) {
+            $orderBy = $this->input['order'][0]['column'];
+            if (isset($this->crud->columns[(int) $orderBy]['name'])) {
+                return [$this->crud->columns[(int) $orderBy]['name'], $this->input['order'][0]['dir']];
+            }
+        }
 
-        return $dataTable->make();
+        return [null, null];
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: Valid but icky :(

File: src/resources/lang/nl/crud.php [hunk 7]
```diff
@@ -56,7 +56,7 @@
         'delete_confirmation_title'                   => 'Item verwijderd',
```
Comment (lloy0076) [inline src/resources/lang/nl/crud.php]: Where did these all come from? (if these are a new lang file, shouldn't they be a separate PR?)

**Comment by tabacitu** [comment]
Hi guys,

@indra1  - thanks a lot for the start, it's been helpful.
@lloy0076 - thanks a lot for the code review, I think I'm the only one who appreciates it but please know I DO :-)

I've finally made A LOT of progress with this - it's been a HUGE endeavour but I think it's ready to use. Here are a few things I'm proud about, or concerned about:

1. For ```select``` and ```select_multiple``` columns, AjaxDataTable now searches within the connected entities. This was a big one :-) This was the expected behaviour, I hope you guys agree.

2. For ```email```, ```date```, ```datetime```, ```text``` columns there's a standard search logic already coded.

3. For ANY other column types (including the ones above or custom column types) you now have the ability to specify a custom search logic. Ex:

```php
        $this->crud->addColumn([
                                'name' => 'slug_or_title',
                                'label' => 'Title',
                                'searchLogic' => function ($query, $column, $searchTerm) {
                                    $query->orWhere('title', 'like', '%'.$searchTerm.'%');
                                }
                            ]);
```

4. ```model_function``` and ```model_function_attribute``` columns now work - they show the right thing in the table, but they WILL NOT be searchable. They need custom search logic, but this should be self-evident for most - there's no way for SQL to search within your PHP.

Downsides:

A. ```text``` columns that actually point to an accessor WILL work in the table, but WILL NOT work in search. They will trigger an error.

B. Because of the above (A), I think we need to qualify this as a breaking change...

What do you guys think? Can anybody test this further or review it? Any ideas on the above?

Cheers!

**Comment by indra1** [comment]
@tabacitu looks amazing. I will start test tomorrow first thing in the morning.

**Comment by Sobak** [comment]
@tabacitu quick suggestion: changes on `composer.json` seem to be unrelated to the PR itself, I think it's better to keep previous indentation in order to reduce diff size.

Can't wait for this to be merged, especially because of #357 being fixed! Any status update on that?

**Comment by adriandmitroca** [comment]
You may find this helpful, but I've found another issue. You can't define multiple `select` columns inside one CRUD, because `name` values of every column override itself - and AjaxDataTables use those values for `SELECT` query, so you can't use anything else than a foreign key column.

It would be extremely useful to be able to set an alias, that would be used either for rendering interface and inside SQL query.

**Comment by jkudish** [comment]
What's the best way for me to test this? I checked out this branch, and ran the publish command for this package, but as far as I could tell it was still loading the old code (loading all rows at once rather then doing an ajax call), am I missing some config?

Let me know! Would be great to test this as I have a project that is struggling with the current data tables.

Also any ETA on the 3.3 release? Was a bit disappointed (though totally understanding!) to see the announcement in the newsletter a few weeks back and then realizing it wasn't release yet ;)

**Comment by tabacitu** [comment]
@jkudish the best way to test this branch is to use the version "dev-ajax-datatables" in your composer.json, then run ```composer update```. In the version you tested you still had to enable the Ajax functionality with ```$this->crud->enableAjaxTable()``` but I just pushed an update where that is the default now. 

ETA by the end of the week! Yup, it pains me it's not released too. It's something I can't wait to have released, because it fixes so many issues. I think we'll just end up releasing every new feature we have so far this week, then break the release cycle a bit with another version in 1-3 months, instead of the usual 6 months. It looks like we still have a lot of work on a few breaking features and I don't want to hold up these finished features for those.

@adriandmitroca - thank you! This is a problem I've heard once or twice before. Alias sounds like a very good idea! Brainstorming - maybe a better name for that would be "key"? I'm wondering if this would be more intuitive. So basically when you define a column, you'd say:
```diff
        $this->crud->addColumn([
+           'key' => 'text_2',
            'name' => 'text',
            'label' => 'Text 2',
            'type' => 'text'
        ]); // add a text column, at the end of the stack
```

If you want the $columns array to use a custom key, instead of the name. This way, you'd be able to use the same column name a bunch of times. This is what you meant, right?

**Comment by tabacitu** [comment]
BOOM! Merged into ```upgrade```. So please use ```dev-upgrade``` in your ```composer.json``` instead of ```dev-ajax-datatables```.

:-) Guys, I'm pretty happy with this branch, I've been using it for a while now with very little issues - that have now all been fixed - so I merged it into the ```upgrade``` branch, which I hope to launch this week. 

Big news: all tables are now officially datatables :-) Whoop-whoop!

@adriandmitroca your issue is still not fixed, but I've submitted a PR for it. See my comment above please.

Cheers! Thanks a lot for the feedback on this one, it's a big big change.

**Comment by adriandmitroca** [comment]
Such a productive day, @tabacitu! Good work!

About the `key` proposal, I think `alias` is much more accurate as `key` sounds like required parameter which is definitely not in this case.

What do you think?

**Comment by MarcosBL** [comment]
Incredible update, will test ASAP

+1 here for the `alias` name, at least is what I'm used to think of in this kind of situations

Congrats again @tabacitu , great step forward !

**Comment by jkudish** [comment]
Nice work @tabacitu 👍

**Comment by tabacitu** [comment]
BOOM! Done, released, please use the [upgrade guide](https://laravel-backpack.readme.io/v3.0/docs/upgrading-crud-32-to-33).

If you have any problems when upgrading, please edit the docs or open a github issue about it.

Cheers!

**Comment by adriandmitroca** [comment]
Love it, seems like I've discovered some critical issue immediately after composer update:

```
> @php artisan package:discover

                                                                               
  [ReflectionException]                                                        
  Class Backpack\Base\Backpack\Base\app\Console\Commands\Install does not exist           
```
